python/paddle/fluid/tests/unittests/test_matmul_op_with_head.py

```diff
@@ -0,0 +1,161 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import unittest
+import numpy as np
+from op_test import OpTest
+
+
+def generate_compatible_shapes_mul_head(dim_X, dim_Y, transpose_X, transpose_Y):
+    BATCH_SIZE = 2
+    M = 3
+    N = 4
+    K = 24
+    if (dim_X == 1 and transpose_X) or (dim_Y == 1 and transpose_Y):
+        K = 1
+    if dim_X == 1:
+        if transpose_X:
+            shape_X = [M]
+        else:
+            shape_X = [K]
+    if dim_Y == 1:
+        if transpose_Y:
+            shape_Y = [N]
+        else:
+            shape_Y = [K]
+    if dim_X >= 2:
+        if transpose_X:
+            shape_X = [K, M]
+        else:
+            shape_X = [M, K]
+    if dim_X == 3:
+        shape_X = [BATCH_SIZE] + shape_X
+    if dim_Y >= 2:
+        if transpose_Y:
+            shape_Y = [N, K]
+        else:
+            shape_Y = [K, N]
+    if dim_Y == 3:
+        shape_Y = [BATCH_SIZE] + shape_Y
+    return shape_X, shape_Y
+
+
+def matmul_head(X, Y, head_number=1):
+    x = []
+    y = []
+    z = []
+    sub_x_width = X.shape[-1] // head_number
+    sub_y_height = Y.shape[-2] // head_number
+    if np.ndim(X) == 2:
+        for i in range(0, head_number):
+            x.append(X[:, i * sub_x_width:i * sub_x_width + sub_x_width])
+            y.append(Y[i * sub_y_height:i * sub_y_height + sub_y_height, :])
+        for i in range(0, head_number):
+            z.append(np.matmul(x[i], y[i]))
+        Z = np.concatenate((z), axis=1)
+
+    elif np.ndim(X) == 3:
+        for i in range(0, head_number):
+            x.append(X[:, :, i * sub_x_width:i * sub_x_width + sub_x_width])
+            y.append(Y[:, i * sub_y_height:i * sub_y_height + sub_y_height, :])
+        for i in range(0, head_number):
+            z.append(np.matmul(x[i], y[i]))
+        Z = np.concatenate((z), axis=2)
+    else:
+        print("ERROR: Not supported dimension")
+
+    return Z
+
+
+def reference_matmul_mul_head(X,
+                              Y,
+                              head_number=1,
+                              transpose_X=False,
+                              transpose_Y=False):
+    """Reference forward implementation using np.matmul."""
+    # np.matmul does not support the transpose flags, so we manually
+    # transpose X and Y appropriately.
+    if transpose_X:
+        if X.ndim == 1:
+            X = X.reshape((X.size, 1))
```

Comment:
Is it necessary to reshape 1-dimensional arrays?

Reply:
Thanks for your comments. Here I re-used the original python/paddle/fluid/tests/unittests/test_matmul_op.py code.

Reply:
OK, but please consider writing better code than in the original test :-)

Reply:
Feel free also to improve the original test.

Reply:
I know that from the scholar point of view a 1-dimensional array can be transposed, but that is perfectly handled by numpy and you do not need to do that manually.
As for scalar outputs, the if branch in the line 108 (https://github.com/PaddlePaddle/Paddle/pull/18570/files#diff-5a8f116278fbc29c2462c4827a5e75e6R108) is written exactly to handle the case of 0 dimentional outputs.

Reply:
OK. will remove the reshaping of 1-dim array.